Introduce SockMan ("lite"): low-level socket handling for HTTP

pinheadmz opened the pull request:
Introduces a new low-level socket manager `SockMan` as an abstract class with virtual functions for implementing higher-level networking protocols like HTTP. This is the next step in https://github.com/bitcoin/bitcoin/issues/31194

This is a minimal, alternative version of #30988 ("Split CConnman") without any changes to working code (P2P is not affected). It adds a stripped-down version of the `SockMan` introduced in that pull request that implements only what is needed for the HTTP server implemented in #32061 (i.e. no I2P stuff and for now, no outbound connection stuff). Exclusions from the original `SockMan` pull request can be checked with:

```
git diff vasild/sockman \
src/common/sockman.h \
src/common/sockman.cpp
```

The commit order has been flipped quite a bit because the original PR incrementally pulls logic out of `net` wheras this PR builds a new system from the bottom-up. Otherwise I tried to keep all the `SockMan` code in order so reviewers of the original PR would be familiar with it.

It also adds unit tests by introducing a `SocketTestingSetup` which mocks network sockets by returning a queue of `DynSock` from `CreateSock()`. HTTP server tests in #32061 will be rebased on this framework as well.

vasild committed:
style: modernize the style of SockMan::AcceptConnection()

pinheadmz committed:
SockMan: introduce class and implement binding to listening socket

Introduce a new low-level socket managing class `SockMan`.

BindListenPort() is copied from CConnMan in net.cpp and will
be modernized in the next commit.

Unit-test it with a new class `SocketTestingSetup` which mocks
`CreateSock()` and will enable mock client I/O in future commits.

`SockMan` and `SocketTestingSetup` are designed to be generic and
reusbale for higher-level network protocol implementation and testing.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

pinheadmz committed:
style: modernize the style of SockMan::BindListenPort()

It was copied verbatim from `CConnman::BindListenPort()` in the previous
commit. Modernize its variables and style and log the error messages
from the caller. Also categorize the informative messages to the "net"
category because they are quite specific to the networking layer.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

pinheadmz committed:
SockMan: implement and test AcceptConnection()

AcceptConnection() is mostly copied from CConmann in net.cpp
and will be modernized in the following commit.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

pinheadmz committed:
SockMan: generate sequential Ids for each newly accepted connection

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

pinheadmz committed:
SockMan: start an I/O loop in a new thread and accept connections

Socket handling methods are copied from CConnMan:

`CConnman::GenerateWaitSockets()` goes to
`SockMan::GenerateWaitSockets()`.

`CConnman::ThreadSocketHandler()` and
`CConnman::SocketHandler()` are combined into
`SockMan::ThreadSocketHandler()`.

`CConnman::SocketHandlerListening()` goes to
`SockMan::SocketHandlerListening()`.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

pinheadmz committed:
SockMan: handle connected sockets: read data from socket

`CConnman::SocketHandlerConnected()` copied to
`SockMan::SocketHandlerConnected()`.

Testing this requires adding a new feature to the SocketTestingSetup,
inserting a "request" payload into the mock client that connects
to us.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

pinheadmz committed:
SockMan: handle connected sockets: write data to socket

Sockets-touching bits from `CConnman::SocketSendData()` copied to
`SockMan::SendBytes()`.

Testing this requires adding a new feature to the SocketTestingSetup,
returning the DynSock I/O pipes from the mock socket so the received
data can be checked.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

pinheadmz committed:
SockMan: dispatch cyclical events from I/O loop

Copy from some parts of `CConnman::SocketHandlerConnected()` and
`CConnman::ThreadSocketHandler()` to:
`EventIOLoopCompletedForOne(id)` and `EventIOLoopCompletedForAll()`.

Co-authored-by: Vasil Dimov <vd@FreeBSD.org>

Sjors commented:
I was able to switch https://github.com/Sjors/bitcoin/pull/50 to use this instead of https://github.com/bitcoin/bitcoin/pull/30988 without have to change anything. So it makes sense to me to focus on this PR first, as it's smaller. Thoughts @vasild?

w0xlt reviewed (commented):
Concept ACK.
Are the functions in this PR to replace Connam functions like `CConnman::BindListenPort`, `CConnman::InitBinds`, etc...?

pinheadmz commented:
> replace Connman functions

In this PR they are just protocol-agnostic copies of those ConnMan functions. Just the socket stuff without the Bitcoin stuff. We could still proceed with the reactor in #30988 after this is merged and actually replace the socket stuff in ConnMan with SockMan but this PR is minimized with the focus being the socket needs of the HTTP server in #32061

vasild commented on src/common/sockman.cpp:
```diff
@@ -0,0 +1,386 @@
+    return m_connected.erase(id) > 0;
+}
+
+ssize_t SockMan::SendBytes(Id id,
```
_3f796b69c8 `SockMan: handle connected sockets: write data to socket`_

In the commit message: `s/recevied/received/`

vasild reviewed (commented):
Approach ACK 5f7941187a12f6d1d180ee29b72b2a5ee7a578b8

pinheadmz committed:
[move-only] Make GetBindAddress() callable from outside net.cpp

pinheadmz reviewed (commented):
Rebase to address review by @vasild THANKS!

vasild reviewed (approved):
ACK 3e7abceecfd790bc0887f647d3f731328e19810f

jonatack commented on src/test/util/setup_common.h:
```diff
@@ -17,6 +17,7 @@
 #include <pubkey.h>
 #include <stdexcept>
 #include <test/util/random.h>
+#include <test/util/net.h>
```
535daaf15fd754335116d17833a45261cdff4e93 nit, sort

jonatack commented on src/test/sockman_tests.cpp:
```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021-2022 The Bitcoin Core developers
```
535daaf15fd754335116d17833a45261cdff4e93

```suggestion
// Copyright (c) 2025-present The Bitcoin Core developers
```

jonatack reviewed (commented):
Began reviewing, then realized that I was reviewing the same code twice with the two "modernize" commits following "original" commits.

Propose to either (1) make clear in the commit message where the code is being pulled from, and that it will be re-styled afterward, or (2) squash the modernization commits into the previous ones. Thereby easing life a bit for your reviewers :)

pinheadmz reviewed (commented):
3e7abceecf -> 598bee6bd5

Address nits from @jonatack including commit messages that indicate code is copied from CConnman and will be modernized in following commits.

hodlinator commented on src/test/util/setup_common.cpp:
```diff
@@ -601,6 +601,42 @@ void TestChain100Setup::MockMempoolMinFee(const CFeeRate& target_feerate)
+
+    // Create the Mock Connected Socket that represents a client.
+    // It needs I/O pipes but its queue can remain empty
+    std::unique_ptr<DynSock> connected_socket{std::make_unique<DynSock>(connected_socket_pipes, std::make_shared<DynSock::Queue>())};
```
Managed to remove `shared_ptr` usage internally in `DynSock`, see commit 7e894f39c41e2ba332aafc5879d7d7380cbcf2fd from my suggestions branch.

pinheadmz replied:
I'm going to keep the `shared_ptr` advice under consideration but didn't apply it to this rebase.

hodlinator commented on src/test/sockman_tests.cpp:
```diff
@@ -0,0 +1,152 @@
+    std::vector<uint8_t> actually_received(expected_response_size);
+    while (!std::ranges::equal(actually_received, sockman.m_respond)) {
+        // Read the data received by the mock socket
+        ssize_t bytes_read = pipes->send.GetBytes((void *)actually_received.data(), expected_response_size);
```
Implemented simulating partial transfer of stream in 6a74e1e9a7393ad51d7a1590384c8a6771b04972 (requires parent commit).

pinheadmz replied:
see above re: packets and chunks

hodlinator commented on src/common/sockman.h:
```diff
@@ -0,0 +1,322 @@
+     * @retval false The connection was refused at the higher level, so the
+     * associated socket and id should be discarded by `SockMan`.
+     */
+    virtual bool EventNewConnectionAccepted(Id id,
+                                            const CService& me,
+                                            const CService& them) = 0;
```
Could add an `EventConnectionClosed()` to mirror this? See 0ca4fac1009017cddff9864ec9cac73d613d67f4.

pinheadmz replied:
No harm in this I suppose but it's not really needed by the HTTP server, we already log client disconnections when calling `CloseConnection()`

hodlinator reviewed (commented):
Concept ACK 598bee6bd590757565d2564ae86cf46b5eea4399

### Concept

Agree with exploring this avenue of introducing `SockMan` without touching `CConnman`. Given @theuni's comment in https://github.com/bitcoin/bitcoin/pull/30988#issuecomment-2702063374, I'm still curious how he would suggest improving it as far as event-loop/send-receive/threading goes.

### Commit approach

Think it's fine so far to pair the commits (copy+modernize, ...). It enables reviewers to review the copying and then diffing over 2 commits at once to see what changed in combination.

### Suggestions branch

https://github.com/bitcoin/bitcoin/compare/598bee6bd590757565d2564ae86cf46b5eea4399...hodlinator:bitcoin:pr/32747_suggestions

Ordered my suggestion commits starting from:
* Trivial - Example: use `std::byte` (44bf80cd90b573d739b3519810449eab6a109992).
* Slightly controversial - Example: decreased `shared_ptr` mentioned ~above~below.
* More arbitrary - Example: rename `ConnectionSockets` => `ConnectionSocket` (650ff1d0762d05b9ea34d527840c5e6c7762afce), probably there are reasons I don't know for the naming.

### Hodlinator's `shared_ptr` corner

`shared_ptr` signals heavily ambiguous ownership. It seems like if there is a socket manager, it should be the one controlling the lifetime of all of them. But since the socket code is already so `shared_ptr`-heavy, it's more like a global network API with some objects with value-like semantics thrown in. *And maybe I can learn to live with that for now.* Few socket operations entail looping over a `vector` of them anyway, which weakens the argument for centralizing ownership. Wrapping them in `shared_ptr` means they can contain inlined data that is "pinned" to that place in memory with decreased risk of being moved and causing issues (even though that could be trivially worked around through internal pointer indirection).

pinheadmz commented on src/test/sockman_tests.cpp:
```diff
@@ -0,0 +1,152 @@
+        EXCLUSIVE_LOCKS_REQUIRED(!m_received_mutex)
+        {
+            LOCK(m_received_mutex);
+            m_received[id].assign(data.begin(), data.end());
```
This is cool but I don't think it's necessary for this PR. In https://github.com/bitcoin/bitcoin/pull/32061 there is more robust testing with real sockets (not `DynSock`) and I think the combining of chunks of data from separate packets is more appropriate one level up (for example in this HTTP server). But I'm also open to integrating this kind of test in the current sockman PR if you think its important

pinheadmz reviewed (commented):
Rebase to e007e1b57d addresses excellent thorough review from @hodlinator (THANKS!). Most significant changes involve casting arguments to socket functions (see #33378) and cleaning up protected/private membership of `SockMan`. I have locally rebased #32061 on these changes and everything still passes ;-) I will update that PR soon based on these changes.

vasild reviewed (approved):
ACK e007e1b57d5d42c2a8d932d5b91eec8a3ca76e14

l0rinc reviewed (commented):
I'm not immersed in this to give a meaningful ack yet, but please see a few of my comments.

My main observation is that we're introducing tests and new code at the same time and I find it hard to follow what the relation between the two is - how do I tell if they're doing what they're supposed to.
Given that this is meant to replace libevent, I wonder if we could add either a wrapper to libevent and cover it with tests as a first step (could be done in a separate standalone PR as well) and slowly replace the methods we rely on from libevent with our own implementation while the tests keep passing and as a last step delete libevent abstraction completely. That would reassure me personally that the new code isn't just dead weight we're adding to the code, but that we're slowly strangling out a dependency with small focused changes covered with tests where I can debug both the old or new behavior and make absolutely sure they behave in the same way.
 
I would also prefer new code to be fully covered with tests, especially since none of the error cases and ipv6 part seem to be.

pinheadmz committed:
Modernize GetBindAddress()

Replace the C-style casting with C++ reinterpret_cast

pinheadmz reviewed (commented):
Rebase to 03ce73a820b485fbdbb854d9d3947ff1e6d87880 address conflicts with master, as well as silent conflict from https://github.com/bitcoin/bitcoin/pull/33378. Address feedback from @l0rinc THANKS -- mostly improving "modernization" of the code borrowed from `Connman`.

There is a bigger question about the approach in this PR, which adds code that is dead for now but required for https://github.com/bitcoin/bitcoin/pull/32061. I am going to explore other ways to introduce this code more incrementally, or maybe we can start off by writing `SockMan` as a libevent wrapper, and then replace libevent with our own code once the libevent behavior is asserted by tests.

vasild reviewed (approved):
ACK 03ce73a820b485fbdbb854d9d3947ff1e6d87880

theuni commented:
My $.02 and repeating what I've said elsewhere... I still just don't think that this is the correct level of abstraction. IMO socket-based protocols are different enough to warrant their own logic. Either that, or the abstraction would need to provide a bazillion options and flags, ala libevent/libev/libuv/libkj/etc.

For ex: presumably in order to remain generic, `SendBytes` may "succeed" at sending only one byte of many. Meaning that implementations will have to maintain their own send buffers, and presumably a thread that does a write/sleep/write/sleep dance. Imo that alone kinda defeats the purpose of this abstraction.

`IOReadiness` and `Sock` already implement a good chunk of the low-level handling. Sure, a custom event-loop per-protocol isn't exactly trivial, but it's a better result imo than a master-of-none abstraction.

I'd rather see this and a custom HTTP server collapsed together into a tight implementation. If another another socket-based protocol is implemented in the future, that would be a good time to abstract away the commonalities. But the fact that CConnman isn't a good a good candidate for abstraction is a hint to me that the next one won't be either.

pinheadmz commented:
rebased on master to fix conflict but also converted to draft.

The plan is now to integrate this I/O loop directly with the new http server. So I will rebase #32061 on this branch to include the feedback the socket stuff has received, and then combine this sockman with the http server and remove the abstraction.
